File: gcj/2012/QC/C.scala
import scala.io.Source

// val input = Source.fromFile("C-small-practice.in")
val input = Source.fromFile("C-large-practice.in")
// val input = Source.fromFile("sample.in")

val divisors:Stream[Int] = 10 #:: divisors.map(_ * 10)

val answers = input.getLines.toList.tail.map(line => {
		val ab = line.split(" ").map(_.toInt)
		count(ab(0), ab(1))
	})

answers.zipWithIndex.foreach(z => println("Case #" + (z._2 + 1) + ": " + z._1))

def count(a: Int, b: Int) = {
	val ret = (a to b).map(n => {
			val order = math.pow(10, n.toString.length).toInt
			val recycled = divisors.takeWhile(_ <= b)
						.map(d => (n % d) * (order / d) + (n / d))
						.toSet

			recycled.count(x => n < x && x <= b)
		}).sum
	// println(a + ", " + b + " done!")
	ret
}

File: gcj/2012/QC/C.java
import java.util.*;
import java.io.*;

public class C {
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new FileReader("C-large-practice.in"));
		br.readLine();
		String l;
		int i = 1;
		while((l = br.readLine()) != null) {
			String[] sa = l.split(" ");
			int n = count(Integer.parseInt(sa[0]), Integer.parseInt(sa[1]));
			System.out.println("Case #" + i + ": " + n);
			i++;
		}
		br.close();
	}

	public static int count(int a, int b) {
		int ret = 0;
		for (int i = a; i < b; i++) {
			Set<Integer> set = new HashSet<Integer>();
			int order = (int)Math.pow(10, Integer.toString(i).length());
			for (int j = 10; j <= b; j *= 10) {
				int recycled = (i % j) * (order / j) + (i / j);
				if (i < recycled && recycled <= b) {
					set.add(recycled);
				}
			}
			ret += set.size();
		}
		return ret;
	}
}
